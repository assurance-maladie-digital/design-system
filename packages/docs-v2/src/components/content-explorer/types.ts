export interface TreeviewItem {
    label: string;
    path: string;
    icon?: string;
    children?: TreeviewItem[]
}