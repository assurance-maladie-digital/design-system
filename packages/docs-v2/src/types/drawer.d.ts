interface PageItem {
	title: string;
	icon?: string;
	href?: string;
	to?: string;
}

export interface DrawerItem {
	title: string;
	icon: string;
	items: PageItem[];
}
