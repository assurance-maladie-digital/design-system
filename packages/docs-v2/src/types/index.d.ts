import { IndexedObject } from '@cnamts/vue-dot/src/types';

export interface ApiProp {
	name: string;
	type: string | string[];
	description: string;
	required?: boolean;
	default?: string | number | boolean | null | undefined;
	signature: string;
	value: string;
}

type ItemValue = Pick<ApiProp, 'name' | 'type' | 'default' | 'required' | 'description'>;
type ItemDescription = Pick<ApiProp, 'name' | 'description'>;
type ItemEvent = Pick<ApiProp, 'name' | 'description' | 'value'>;
type ItemFunction = Pick<ApiProp, 'name' | 'signature' | 'description'>;
type ItemModifier = Pick<ApiProp, 'name' | 'type' | 'description'>;
type ItemArgument = Pick<ApiProp, 'type' | 'description'>;

interface ApiHeaderList {
	options: ItemValue[];
	slots: ItemDescription[];
	props: ItemValue[];
	events: ItemEvent[];
	functions: ItemFunction[];
	modifiers: ItemModifier[];
	argument: ItemArgument[];
}

export type HeaderList = keyof ApiHeaderList;
export type ApiHeaders = Partial<ApiHeaderList>;

export type Api = IndexedObject<ApiHeaders>;
