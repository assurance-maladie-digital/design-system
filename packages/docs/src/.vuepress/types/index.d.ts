import { IndexedObject } from "@cnamts/vue-dot/src/types";

export interface ItemHeader {
	label: string;
	value: string;
	class: string;
}

interface Item {
	[key: string]: string | boolean | undefined;
	value?: string;
	description: string;
	name: string;
	type: string;
	required?: boolean;
	signature?: string;
	options?: string;
	snippet?: string;
	props?: string;
}

type ApiItem = IndexedObject<Item[]>
export type ApiItems = IndexedObject<ApiItem>;

interface SideBarLink {
	title: string;
	to: string;
}

export interface SidebarItem {
	title: string;
	group: string;
	items: SideBarLink[];
}
