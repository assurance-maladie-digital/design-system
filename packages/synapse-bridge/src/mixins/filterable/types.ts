import type { ChipItem } from '@/elements/ChipList/types';

export interface FilterItem {
  name: string
  value?: unknown
  formatChip?: (value: any) => ChipItem[]
  chipOverflowLimit?: number
  title?: string
  [key: string]: unknown
}

export { type ChipItem };
