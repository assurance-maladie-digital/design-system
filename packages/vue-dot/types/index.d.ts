import { VueConstructor, PluginFunction } from 'vue';
import { Framework } from 'vuetify/types';

// Types for the library usage
// eg. Vue.use(VueDot, options)

export interface Icons {
	[key: string]: string;
}

export interface Theme {
	icons: Icons;
}

export interface LocalStorageControl {
	version?: number;
	expiration?: number;
	prefix?: string;
}

export interface VueDotOptions {
	theme?: Theme;
	localStorageControl: LocalStorageControl;
}

export interface VueDotPlugin {
	install: PluginFunction<VueDotOptions>;
}

declare const VueDot: VueDotPlugin;

declare global {
	interface Window {
		Vue?: VueConstructor;
	}

	/** @see https://github.com/microsoft/TypeScript/issues/45612 */
	interface Navigator {
		/* eslint-disable @typescript-eslint/no-explicit-any */
		msSaveBlob(blob: any, defaultName?: string): boolean;
		msSaveOrOpenBlob(blob: any, defaultName?: string): boolean;
	}
}

declare module 'vue/types/vue' {
	interface Vue {
		$vd: VueDotOptions;
		$vuetify: Framework;
	}
}

export default VueDot;
