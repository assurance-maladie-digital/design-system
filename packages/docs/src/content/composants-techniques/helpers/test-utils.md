---
title: testUtils
description: Le helper `testUtils` contient un ensemble de fonctions utilitaires pour les tests unitaires.
---

<doc-tabs>

<doc-tab-item label="Utilisation">

<doc-alert type="info">

Le helper `testUtils` exporte également toutes les fonctions du package [Vue Test Utils](https://vue-test-utils.vuejs.org/).

</doc-alert>

### applyGlobalConfig

La fonction `applyGlobalConfig` permet d’appliquer la configuration globale pour les tests, elle est utilisée dans le fichier `setup.ts`.

```ts
import { applyGlobalConfig } from '@cnamts/vue-dot/src/helpers/testUtils';

applyGlobalConfig();
```

### installGlobalPlugins

La fonction `installGlobalPlugins` permet d’installer les plugins [Vue Input Facade](https://ronaldjerez.github.io/vue-input-facade/), [Vue Meta](https://vue-meta.nuxtjs.org/), [Vuex](https://vuex.vuejs.org/), [Vuetify](https://v2.vuetifyjs.com/) et Vue Dot pour les tests.

```ts
import { createLocalVue, installGlobalPlugins } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installGlobalPlugins(localVue);
```

### installRouter

La fonction `installRouter` permet d’installer [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createLocalVue, installRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const localVue = createLocalVue();

installRouter(localVue);
```

### createRouter

La fonction `createRouter` permet de créer une instance de [Vue Router](https://v3.router.vuejs.org/fr/) pour les tests.

```ts
import { createRouter } from '@cnamts/vue-dot/src/helpers/testUtils';

const routes = [];

const router = createRouter({
	routes
});
```

### createStore

La fonction `createStore` permet de créer une instance de [Vuex](https://vuex.vuejs.org/) pour les tests.

```ts
import { createStore } from '@cnamts/vue-dot/src/helpers/testUtils';

const store = createStore({
	state: {}
});
```

### createVuetifyInstance

La fonction `createVuetifyInstance` permet de créer une instance de [Vuetify](https://v2.vuetifyjs.com/) pour les tests.

```ts
import { createVuetifyInstance } from '@cnamts/vue-dot/src/helpers/testUtils';

const vuetify = createVuetifyInstance();
```

### mockVFormRef

La fonction `mockVFormRef` permet de mocker la référence d’un composant [VForm](https://v2.vuetifyjs.com/en/components/forms/) pour les tests.

```ts
import { mockVFormRef } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	$refs: {
		form: mockVFormRef(true)
	}
};
```

### mockTranslations

La fonction `mockTranslations` permet de mocker les traductions pour les tests.

```ts
import { mockTranslations } from '@cnamts/vue-dot/src/helpers/testUtils';

const mocks = {
	...mockTranslations<string | string[]>({
		'path.to.title': 'Titre',
		'path.to.items': [
			'Élément 1',
			'Élément 2',
			'Élément 3'
		]
	})
};
```

</doc-tab-item>

<doc-tab-item label="API">
<doc-api name="helpers/test-utils"></doc-api>
</doc-tab-item>

</doc-tabs>
