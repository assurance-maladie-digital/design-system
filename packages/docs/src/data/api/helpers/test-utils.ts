import { Api } from '~/types';

export const api: Api = {
	applyGlobalConfig: {
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installGlobalPlugins: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	installRouter: {
		arguments: [
			{
				name: 'localVue',
				description: 'Local Vue instance',
				type: 'Vue',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	createRouter: {
		arguments: [
			{
				name: 'options',
				description: 'Router options',
				type: 'RouterOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'VueRouter',
			},
		],
	},
	createStore: {
		arguments: [
			{
				name: 'options',
				description: 'Store options',
				type: 'StoreOptions',
				required: false,
			},
		],
		returnValue: [
			{
				type: 'Store',
			},
		],
	},
	createVuetifyInstance: {
		arguments: [
			{
				name: 'preset',
				description: 'Vuetify preset',
				type: 'Partial<UserVuetifyPreset>',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'Vuetify',
			},
		],
	},
	mockVFormRef: {
		arguments: [
			{
				name: 'isValid',
				description: 'Is valid',
				type: 'boolean',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
	mockTranslations: {
		arguments: [
			{
				name: 'translations',
				description: 'Translations',
				type: 'IndexedObject',
				required: true,
			},
		],
		returnValue: [
			{
				type: 'void',
			},
		],
	},
};
