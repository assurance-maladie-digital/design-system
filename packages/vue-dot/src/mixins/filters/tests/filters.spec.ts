import Vue from 'vue';
import { mount, Wrapper } from '@vue/test-utils';

import { FilterMixin } from '../';
import { FilterItem, ChipItem } from '../types';

interface TestComponent extends Vue {
	isMobile: boolean;
	applyFunction(): unknown;
	resetFilter(): unknown;
	removeChip(event: ChipItem): unknown;
	closeSidebar(): unknown;
	filters: FilterItem[];
	onChange(event: unknown, filters: FilterItem): void;
	removeAccents(str: string): string;
	getChipValue(event: unknown): unknown;
	formatting(value: unknown): string | null | undefined;
}

const filterExemple: FilterItem = {
	name: 'prenom',
	label: 'Prénom',
	chips: [],
	value: '',
	clearAfterValidate: true,
	limited: false,
	splitted: false,
	description: '',
	showAll: false,
	icon: '',
	selection: [],
	defaultValue: undefined,
	style: ''
};
const limitedFilterExemple: FilterItem = {
	...filterExemple,
	limited: true
};
const splittedFilterExemple: FilterItem = {
	...filterExemple,
	splitted: true
};
const defaultValueExemple: FilterItem = {
	...filterExemple,
	defaultValue: 'defaultValue'
};
const chipItemEvent: ChipItem = {
	text: 'example',
	value: 'event'
};

/** Create the test component */
function createTestComponent() {
	return Vue.component('TestComponent', {
		mixins: [
			FilterMixin
		],
		data() {
			return {
				filters: [
					filterExemple,
					limitedFilterExemple,
					splittedFilterExemple,
					defaultValueExemple
				]
			};
		},
		template: '<div />'
	});
}

describe('filters', () => {
	it('test filter mixin', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		expect(wrapper.vm.filters).toMatchSnapshot();
	});

	it('should not add a new chip to the filter when the event is undefined', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.onChange(undefined, filterExemple);

		expect(wrapper.vm.filters[0].chips).toHaveLength(0);
	});

	it('should add a new chip to the filter when onChange is called', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.onChange('example', filterExemple);

		expect(wrapper.vm.filters[0].chips).toContainEqual({
			text: 'example',
			value: 'example'
		});
	});

	it('should not add a new chip to the filter when the chip already exists', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.onChange('example', filterExemple);
		wrapper.vm.onChange('example', filterExemple);

		expect(wrapper.vm.filters[0].chips).toHaveLength(1);
	});

	it('should return a string value when onChange is called with a string argument', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		expect(wrapper.vm.getChipValue('example')).toBe('example');
	});

	it('should return an object value when onChange is called with an object argument', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		expect(wrapper.vm.getChipValue({ foo: 'bar' })).toBe('bar');
	});

	it('should clear the value of the filter when clearAfterValidate is true', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.onChange('foo', filterExemple);

		expect(wrapper.vm.filters[0].value).toBeNull();
	});

	it('should not clear the value of the filter when clearAfterValidate is true and defaultValue is defined', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.onChange('foo', defaultValueExemple);

		expect(wrapper.vm.filters[3].value).toBe('defaultValue');
	});

	it('should emit the onchange-filters event when onChange is called', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.$emit('onchange-filters', 'example');

		wrapper.vm.$nextTick(() => {
			expect(wrapper.emitted('onchange-filters')).toBeTruthy();
		});
	});

	it('should add a new chip to the filter when limited is true', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.onChange('example', wrapper.vm.filters[1]);

		expect(wrapper.vm.filters[1].chips).toContainEqual({
			text: 'example',
			value: 'example'
		});
	});

	it('should split the value into chips when splitted is true', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.onChange({ foo: 'bar', baz: 'qux' }, wrapper.vm.filters[2]);

		expect(wrapper.vm.filters[2].chips).toStrictEqual([
			{ text: 'bar', value: 'bar' },
			{ text: 'qux', value: 'qux' }
		]);
	});

	it('should not split the value into chips when splitted is true and the value is not an object', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.onChange('example', wrapper.vm.filters[2]);

		expect(wrapper.vm.filters[2].chips).toStrictEqual([]);
	});

	it('should return the chip value with formatting function', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.formatting = (event: ChipItem) => event.text.toUpperCase();
		wrapper.vm.onChange('example', wrapper.vm.filters[0]);

		if (wrapper.vm.formatting) {
			expect(wrapper.vm.formatting(chipItemEvent)).toBe('EXAMPLE');
		}
	});

	it('should return the chip value with null formatting function', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.formatting = () => null;
		wrapper.vm.onChange('example', wrapper.vm.filters[0]);

		if (!wrapper.vm.formatting) {
			expect(wrapper.vm.formatting(chipItemEvent)).toBe('example');
		}
	});

	it('should split the value into chips when splitted is true and the value of the object use formatting function', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.formatting = (event: ChipItem) => event.text.toUpperCase();
		wrapper.vm.onChange({ firstName: 'Alice', lastName: 'Smith' }, wrapper.vm.filters[2]);

		expect(wrapper.vm.filters[2].chips).not.toStrictEqual([
			{ text: 'ALICE', value: 'ALICE' },
			{ text: 'SMITH', value: 'SMITH' }
		]);
	});

	it('should return true if the screen is mobile', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent, {
			mocks: {
				$vuetify: {
					breakpoint: {
						smAndDown: true
					}
				}
			}
		});

		expect(wrapper.vm.isMobile).toBe(true);
	});

	it('should remove accents from a string', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		expect(wrapper.vm.removeAccents('áéíóú')).toBe('aeiou');
	});

	it('should emit the close-sidebar event when closeSidebar is called', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.closeSidebar();

		expect(wrapper.emitted('close-sidebar')).toBeTruthy();
	});

	it('should emit the remove-chip event when removeChip is called', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.removeChip(chipItemEvent);

		expect(wrapper.emitted('remove-chip')).toBeTruthy();
	});

	it ('should emit the reset-filter event when resetFilter is called', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.resetFilter();

		expect(wrapper.emitted('reset-filter')).toBeTruthy();
	});

	it('should emit the apply-function event when applyFunction is called', () => {
		const testComponent = createTestComponent();
		const wrapper = mount(testComponent) as Wrapper<TestComponent>;

		wrapper.vm.applyFunction();

		expect(wrapper.emitted('apply-function')).toBeTruthy();
	});
});
