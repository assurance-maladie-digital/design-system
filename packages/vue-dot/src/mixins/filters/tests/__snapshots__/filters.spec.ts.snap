// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters test filter mixin 1`] = `
Array [
  Object {
    "chips": Array [],
    "clearAfterValidate": true,
    "defaultValue": undefined,
    "description": "",
    "icon": "",
    "label": "Prénom",
    "limited": false,
    "name": "prenom",
    "selection": Array [],
    "showAll": false,
    "splited": false,
    "style": "",
    "value": "",
  },
  Object {
    "chips": Array [],
    "clearAfterValidate": true,
    "defaultValue": undefined,
    "description": "",
    "icon": "",
    "label": "Prénom",
    "limited": true,
    "name": "prenom",
    "selection": Array [],
    "showAll": false,
    "splited": false,
    "style": "",
    "value": "",
  },
  Object {
    "chips": Array [],
    "clearAfterValidate": true,
    "defaultValue": undefined,
    "description": "",
    "icon": "",
    "label": "Prénom",
    "limited": false,
    "name": "prenom",
    "selection": Array [],
    "showAll": false,
    "splited": true,
    "style": "",
    "value": "",
  },
  Object {
    "chips": Array [],
    "clearAfterValidate": true,
    "defaultValue": "defaultValue",
    "description": "",
    "icon": "",
    "label": "Prénom",
    "limited": false,
    "name": "prenom",
    "selection": Array [],
    "showAll": false,
    "splited": false,
    "style": "",
    "value": "",
  },
]
`;
